[Merged by Bors] - Builder Specs v0.2.0

realbigsean opened the pull request:
## Issue Addressed

https://github.com/sigp/lighthouse/issues/3091

Extends https://github.com/sigp/lighthouse/pull/3062, adding pre-bellatrix block support on blinded endpoints and allowing the normal proposal flow (local payload construction) on blinded endpoints. This resulted in better fallback logic because the VC will not have to switch endpoints on failure in the BN <> Builder API, the BN can just fallback immediately and without repeating block processing that it shouldn't need to. We can also keep VC fallback from the VC<>BN API's blinded endpoint to full endpoint.

## Proposed Changes

- Pre-bellatrix blocks on blinded endpoints
- Add a new `PayloadCache` to the execution layer
- Better fallback-from-builder logic

## Todos

- [x] Remove VC transition logic
- [x] Add logic to only enable builder flow after Merge transition finalization
- [x] Tests
- [x] Fix metrics
- [x] Rustdocs

macladson committed:
Update HTTP API to include `execution_optimistic` flag in responses

macladson committed:
Update HTTP API to include `execution_optimistic` flag in responses

macladson committed:
Squash previous works

Update HTTP API to include `execution_optimistic` flag in responses

Add `execution_optimistic` to server-sent events

Fix formatting

Fix tests

Fix endpoint comment

Remove redundent calls to `get_block`

Fix route

Return correct `execution_optimistic` values for api queries by state

Updated api client functions

Refactor HTTP API tests

Return 404 when requested root is 0x000..000

Tidy up

Avoid recomputing head

More improvements

Use head when computing status for slot

Squash several commit for 3175

Start adding ChainView

Ripple `ChainView` across code-base

Fix after rebase on #3070

Pass `make lint` with `chain_view`

Remove uses of get_head

Remove uses of `with_head`

Remove use of canonical head

Start combining head and fork choice

Big refactor to further lock joining

Fix to pass `make lint`

Fix consensus/fork_choice tests

Fix http_api tests

Remove current_slot from FC rebuild

Fix lint issues after rebase

Refactor out into recompute_head.rs

Add TaskExecutor to BeaconChainHarness

Thread task_executor into BeaconChain

Use async fork choice

Fix test compile errors

Clear early attester cache sooner

Move canonical_head to own file, fix test deadlock

Get `make lint` working again

Add `genesis_time` to `BeaconChain`

Tidy

Fix test compile issues

Move recompute_head fns together

Spawn EL update tasks

Move EL updates into own function

Tidy

macladson committed:
Fix route

macladson committed:
Return correct `execution_optimistic` values for api queries by state

macladson committed:
Updated api client functions

realbigsean commented:
This doesn't match the semantics of the upcoming changes to the builder API: https://github.com/flashbots/mev-boost/pull/82

But I think it'd still be worth working towards merging this as is, rather than waiting on the stabilization of that spec if we plan to use the payload cache elsewhere. This also gets us in line with the standard API spec by making sure we can provide pre-bellatrix blocks on blinded endpoints.

macladson committed:
Refactor HTTP API tests

macladson committed:
Return 404 when requested root is 0x000..000

macladson committed:
Tidy up

macladson committed:
Avoid recomputing head

macladson committed:
More improvements

macladson committed:
Use head when computing status for slot

paulhauner commented:
There's a conflict here :)

macladson committed:
Rebase on unstable

macladson committed:
Add tests

macladson committed:
Fix conflicts

realbigsean committed:
fix signing

realbigsean committed:
Merge branch 'register-validator-api' of https://github.com/realbigsean/lighthouse into pre-bellatrix-blinded-blocks

realbigsean committed:
make sure post to validator registration endpoint expects no response

realbigsean committed:
allow mock el dead code

realbigsean committed:
allow mock el dead code

realbigsean committed:
wrap the payload header in a bid

realbigsean committed:
fix query path

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

realbigsean committed:
fix bid deserialization

realbigsean committed:
Update beacon_node/http_api/tests/tests.rs

realbigsean committed:
Update consensus/types/src/chain_spec.rs

realbigsean committed:
Update validator_client/src/preparation_service.rs

realbigsean committed:
Update validator_client/src/preparation_service.rs

realbigsean committed:
Update validator_client/src/preparation_service.rs

realbigsean committed:
actually do bitmask operation in test

realbigsean committed:
don't double publish changed keys on epoch boundary

realbigsean committed:
only start the registration service if we have the private transactions proposals flag

realbigsean committed:
publish registrations regardless of how close we are to the merge

realbigsean committed:
fix lints, commented out tests for future web3signer addition

realbigsean committed:
supress lint

realbigsean committed:
Merge branch 'register-validator-api' of https://github.com/realbigsean/lighthouse into pre-bellatrix-blinded-blocks

realbigsean committed:
add timeout to the builder client header endpoint

realbigsean committed:
batched validator registration data

realbigsean committed:
add mock builder

realbigsean committed:
Merge branch 'ex-opt-api' of https://github.com/macladson/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	beacon_node/http_api/Cargo.toml

realbigsean committed:
add builder to api tests

realbigsean committed:
fix signing in test

realbigsean committed:
context dep public

realbigsean committed:
context dep public

paulhauner committed:
Resolve `BeaconChain` compile errors

paulhauner committed:
Revert `common/eth2` to `unstable`

paulhauner committed:
Remove old fork choice functions

paulhauner committed:
Fix compile errors in HTTP tests

paulhauner committed:
Wrap harness in Arc

paulhauner committed:
Start *horrific* async test refactor

paulhauner committed:
More test fixes

realbigsean committed:
logs for debugging

realbigsean committed:
logs for debugging

realbigsean committed:
sign builder bids in mock builder

realbigsean committed:
key payload cache by tree hash root not tx root

realbigsean committed:
key payload cache by tree hash root not tx root

realbigsean committed:
sign builder bids in mock builder

realbigsean committed:
sign builder bids in mock builder

realbigsean committed:
debug logs

realbigsean committed:
fix prev randao in mock builder

realbigsean committed:
fix randao in mock builder

realbigsean committed:
json serde mock builder payloads

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

paulhauner committed:
Make tests *finally* compile with async FC

realbigsean committed:
logging updates

realbigsean committed:
fix typo

paulhauner committed:
Add progress on async block import

paulhauner committed:
Solve `StoreOp` lifetime problem

realbigsean committed:
improve local testing

realbigsean committed:
cago fmt, cargo fix, cargo generate-lockfile

realbigsean committed:
fix ports in tests

paulhauner committed:
Reinstate panda banner

paulhauner committed:
Partially add `PayloadNotifier`

paulhauner committed:
Start Arc-ifying SignedBeaconBlock

paulhauner committed:
Arc-ification of networking crate

paulhauner committed:
Fix http_api, `make lint` passes

paulhauner committed:
Refactor exec payload "prepare payload"

paulhauner committed:
Make block production async (make lint passing)

paulhauner committed:
Fix test compilation errors

paulhauner committed:
Fix justified checkpoint getter

paulhauner committed:
Remove deadlock from block import

paulhauner committed:
Fix `http_api` to pass `make lint`

paulhauner committed:
Fix clippy lint

paulhauner committed:
Update test for finalized_checkpoint event

paulhauner committed:
Fix MSRV issue

paulhauner committed:
Fix failing network tests

paulhauner committed:
Remove HeadSafetyStatus

paulhauner committed:
Remove unused error

paulhauner committed:
Fix compile errors in beacon chain tests

paulhauner committed:
Simplify chain segment processing

paulhauner committed:
Remove some blocking methods

paulhauner committed:
Fix missing `!`

realbigsean committed:
Validate some fields vs a local payload

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	Cargo.lock

realbigsean committed:
fmt and remove reference

paulhauner committed:
Fix failing payload_invalidation tests

paulhauner committed:
Fix store test

paulhauner committed:
Remove commented-out code

paulhauner committed:
Fix weak subjectivity test

paulhauner committed:
Move EL update spawning

paulhauner committed:
Fix http_api compile error

paulhauner committed:
Fix compile error in client builder

paulhauner committed:
Remove genesis alias workaround from EF tests

paulhauner committed:
Remove `block_on` from EL

paulhauner committed:
Remove block_on for blinded blocks in HTTP API

realbigsean committed:
fix pre-TTD block proposals

realbigsean committed:
fix fix

realbigsean committed:
pawan > me

paulhauner committed:
Disable rayon for fork choice tests

paulhauner committed:
Simplify getting for active val count

paulhauner committed:
Make block_on_dangerous respect exit, metrics

paulhauner committed:
Patch and unify status_message

realbigsean committed:
start docs

paulhauner committed:
Move recompute_head into canonical_head

paulhauner committed:
Remove head_proposer_shuffling_decision_root

paulhauner committed:
Arc-ify BeaconSnapshot

paulhauner committed:
Spawn new task for after_finalization

paulhauner committed:
Provide different state root to db migrator

paulhauner committed:
Don't log crit when tokio stops block import

paulhauner committed:
Add back warning for opt head

paulhauner committed:
Remove debug formatting for roots

paulhauner committed:
Raise gossip block log levels

paulhauner committed:
Revert "Remove debug formatting for roots"

This reverts commit 3f141ffd6f910e8f538737cfbb004e43381e98f3.

paulhauner committed:
Add weight logging

paulhauner committed:
Drop txn_lock

paulhauner committed:
Compute state root in spawned function

paulhauner committed:
Use state roots iter directly

realbigsean committed:
start docs

paulhauner committed:
Avoid blocking async on canon head lock

paulhauner committed:
Add fast canonical head

paulhauner committed:
Remove inconsistent head errors

paulhauner committed:
Update fast head earlier

paulhauner committed:
Ensure FFG checkpoints are consistent with head

paulhauner committed:
Do timeout lock for beacon state metrics

paulhauner committed:
Use fast head in http_api

paulhauner committed:
Use fast head in notifier

realbigsean committed:
update docker version

realbigsean committed:
lints

realbigsean committed:
cargo fmt

paulhauner committed:
Split head and FC locks

paulhauner committed:
Get make lint passing

paulhauner committed:
Tidy, add comments

paulhauner committed:
Tidy comments

realbigsean committed:
pin deps

realbigsean committed:
Revert "pin deps"

This reverts commit c41b73e908ab8ce79cfaba9286a48a733c8a2593.

realbigsean committed:
pin deps

realbigsean committed:
Revert "pin deps"

This reverts commit 7dc5d4125f551da8e6fd6d136edb9e6486f0095f.

paulhauner committed:
Switch to `Arc<Snapshot>` for `CachedHead`

paulhauner committed:
Expose fork choice locks

paulhauner committed:
Tidy

paulhauner committed:
Fix test compile errors

paulhauner committed:
Update Cargo.lock

paulhauner committed:
Unify is_optimistic_candidate_block

paulhauner committed:
Fix flipped bool

paulhauner committed:
Revert changes to clippy.toml

paulhauner committed:
Revert changes to beacon_chain Cargo.toml

paulhauner committed:
Refactor head accessor methods

paulhauner committed:
Fix broken store test

paulhauner committed:
Fix state_advance_timer TODO

paulhauner committed:
Refactor `BeaconChain::best_slot`

paulhauner committed:
Fix conflict in `state_advance_timer`

paulhauner committed:
Use blocking task for FC notifier tx

paulhauner committed:
Update opt sync error message

paulhauner committed:
Remove commented-out code from beacon_processor

paulhauner committed:
Add comments to `TaskSpawner`

paulhauner committed:
Reduce duplicate code in sync_methods

paulhauner committed:
Add comment about val count caches

paulhauner committed:
Update block_on_dangerous comments

paulhauner committed:
Remove phantom block_efficiency file

paulhauner committed:
Remove parking lot from ef_tests

paulhauner committed:
Use rayon in ef_tests by default

paulhauner committed:
Remove erroneous log for opt transition block

paulhauner committed:
Tidy

paulhauner committed:
Undo changes to `BeaconChain::import_block`

paulhauner committed:
Reduce diff for attestation production

paulhauner committed:
Reduce chain segment diff

paulhauner committed:
Tidy

paulhauner committed:
Tidy

paulhauner committed:
Tidy

paulhauner committed:
Apply suggestions from code review

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

paulhauner committed:
Use fc write lock for restore

paulhauner committed:
Address review comments from @realbigsean

paulhauner committed:
Apply suggestions from code review

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

paulhauner committed:
Fix bad comment

realbigsean committed:
Merge branch 'nfd-2' of https://github.com/paulhauner/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	Cargo.lock
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/beacon_chain/src/beacon_proposer_cache.rs
	beacon_node/beacon_chain/src/execution_payload.rs
	beacon_node/execution_layer/src/engines.rs
	beacon_node/execution_layer/src/lib.rs
	beacon_node/http_api/src/attester_duties.rs
	beacon_node/http_api/src/lib.rs
	beacon_node/http_api/src/proposer_duties.rs
	beacon_node/http_api/src/state_id.rs
	beacon_node/http_api/tests/tests.rs

realbigsean committed:
cargo lock

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	Cargo.lock
	beacon_node/execution_layer/src/engine_api.rs
	beacon_node/execution_layer/src/engine_api/http.rs
	beacon_node/execution_layer/src/lib.rs
	beacon_node/http_api/Cargo.toml
	beacon_node/src/cli.rs
	beacon_node/src/config.rs

realbigsean committed:
remove duplicate dep

realbigsean committed:
fix compile errors after merging

realbigsean committed:
fix lints in tests

realbigsean committed:
fix tests and handle 204

macladson committed:
Fix tests and apply suggestions

macladson committed:
Fix tests and apply suggestions

macladson committed:
Get blinded blocks where possible

macladson committed:
Get blinded blocks where possible

macladson committed:
Missed one

macladson committed:
Missed one

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into ex-opt-unstable

realbigsean committed:
Merge branch 'ex-opt-unstable' of https://github.com/realbigsean/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/beacon_chain/src/canonical_head.rs
	beacon_node/beacon_chain/src/execution_payload.rs
	beacon_node/beacon_chain/tests/payload_invalidation.rs
	beacon_node/builder_client/Cargo.toml
	beacon_node/builder_client/src/lib.rs
	beacon_node/execution_layer/Cargo.toml
	beacon_node/execution_layer/src/engines.rs
	beacon_node/execution_layer/src/lib.rs
	beacon_node/http_api/Cargo.toml
	beacon_node/http_api/src/block_id.rs
	beacon_node/http_api/src/lib.rs
	beacon_node/http_api/src/proposer_duties.rs
	beacon_node/http_api/src/state_id.rs
	beacon_node/http_api/tests/tests.rs
	beacon_node/network/src/beacon_processor/worker/sync_methods.rs
	beacon_node/src/cli.rs
	beacon_node/src/config.rs
	common/task_executor/src/lib.rs
	common/task_executor/src/metrics.rs
	consensus/types/src/application_domain.rs
	consensus/types/src/beacon_block.rs
	consensus/types/src/beacon_block_body.rs
	consensus/types/src/beacon_state/committee_cache/tests.rs
	consensus/types/src/chain_spec.rs
	validator_client/src/preparation_service.rs
	validator_client/src/signing_method.rs

realbigsean committed:
fix merge issues

realbigsean committed:
lints

realbigsean commented:
Blocked on #3070

realbigsean committed:
enable validator registration web3signer tests

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	Dockerfile
	beacon_node/execution_layer/src/engine_api/json_structures.rs

ciaranmcveigh5 committed:
add flag that enables users to override the builder registration time… (#3340)

* add flag that enables users to override the builder registration timestamp field, required for distributed validators

* change type to option<u64>

* Fix compile issues

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

* add recommended test cases

* mismatched closing delimiter

* cargo fmt

Co-authored-by: realbigsean <seananderson33@GMAIL.com>

macladson committed:
Implement suggestions

macladson committed:
Wrap execution_optimistic in an Option

realbigsean committed:
[builder API] Verify builder signature (#3334)

* verify fork from relay

* verify builder signature and fork

realbigsean committed:
[builder API] Add/update relay timeouts (#3333)

* add/update timeouts

* make sure all requests use timeouts

realbigsean committed:
[builder API] Chain health checks before using builder API (#3356)

* add builder health checks

* remove `requires("builder")` because it makes the `--builder` flag required

* Update chain health checks to an enum

* add flag to disable builder fallback checks

realbigsean committed:
[builder API] Make gas limit configurable, and add per-validator builderAPI usage (#3342)

* make gas limit configurable

* per-validator builder API enabling

* relax fee recipient checks

* update fee recipient test

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into builder-api

 Conflicts:
	beacon_node/beacon_chain/src/beacon_chain.rs
	beacon_node/beacon_chain/src/execution_payload.rs
	beacon_node/beacon_chain/src/test_utils.rs
	beacon_node/execution_layer/Cargo.toml
	beacon_node/execution_layer/src/lib.rs
	beacon_node/execution_layer/src/test_utils/handle_rpc.rs
	beacon_node/execution_layer/src/test_utils/mock_execution_layer.rs
	beacon_node/execution_layer/src/test_utils/mod.rs
	beacon_node/http_api/src/lib.rs
	lcli/Dockerfile
	testing/antithesis/Dockerfile.libvoidstar
	testing/execution_engine_integration/src/test_rig.rs

realbigsean committed:
merge with unstable

macladson committed:
Undo regression

realbigsean committed:
fix tests and block hash check

realbigsean committed:
Merge branch 'ex-opt-api' of https://github.com/macladson/lighthouse into builder-api

 Conflicts:
	beacon_node/http_api/Cargo.toml
	beacon_node/http_api/src/attester_duties.rs
	beacon_node/http_api/src/lib.rs
	beacon_node/http_api/src/proposer_duties.rs
	beacon_node/http_api/src/state_id.rs
	beacon_node/http_api/src/version.rs
	beacon_node/http_api/tests/tests.rs
	common/eth2/src/types.rs

realbigsean committed:
lint

realbigsean committed:
lint

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into builder-api

realbigsean committed:
query v1 blinded blocks endpoint as per beacon apis spec

realbigsean committed:
fix fee recipient condition

realbigsean committed:
add more builder validation tests

realbigsean committed:
fix test

realbigsean committed:
fix cli tests

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	beacon_node/http_api/Cargo.toml
	beacon_node/http_api/src/attester_duties.rs
	beacon_node/http_api/src/block_id.rs
	beacon_node/http_api/src/lib.rs
	beacon_node/http_api/src/proposer_duties.rs
	beacon_node/http_api/src/state_id.rs
	beacon_node/http_api/src/validator_inclusion.rs
	beacon_node/http_api/tests/tests.rs

realbigsean committed:
fix merge with unstable issues

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	beacon_node/beacon_chain/src/chain_config.rs
	beacon_node/http_api/src/lib.rs
	beacon_node/src/cli.rs
	beacon_node/src/config.rs
	consensus/types/src/payload.rs
	lighthouse/tests/validator_client.rs
	validator_client/src/block_service.rs
	validator_client/src/cli.rs
	validator_client/src/config.rs
	validator_client/src/lib.rs

realbigsean committed:
fix merge with unstable

realbigsean commented on common/eth2/src/lib.rs:
```diff
@@ -1291,7 +1291,7 @@ impl BeaconNodeHttpClient {
         verify_randao: Option<bool>,
     ) -> Result<ForkVersionedResponse<BeaconBlock<T, Payload>>, Error> {
-        let mut path = self.eth_path(V2)?;
+        let mut path = self.eth_path(V1)?;
```
Sneaky bugfix, the spec says this should be V1, our beacon HTTP server will accept any version so this may have kept us from working cross client but wouldn't have been an issue with a full lighthouse setup

realbigsean committed:
git commit -am "self-review fixes"

realbigsean committed:
make sure PATCH to validator def updates initialized validators

realbigsean commented:
I have yet to really add docs, but this is ready for review otherwise. I am going to work on docs tomorrow. The issues linked [here](https://github.com/sigp/lighthouse/issues/3091) include a lot of the reasoning for the changes here. If this PR is too large for anyone's appetite let me know and I can work to split it up along the lines of the issues included.

realbigsean committed:
improve logging, separate blinded and full block request metrics for VC, make sure the validator registration service always runs in case the builder proposals flag is not provided but the validator defs file includes the config

michaelsproul commented on .github/workflows/local-testnet.yml:
```diff
@@ -40,13 +40,29 @@ jobs:
 
       - name: Start local testnet
-        run: ./start_local_testnet.sh
+        run: ./start_local_testnet.sh && sleep 60
```
Is the `sleep` necessary?

michaelsproul replied:
(don't mind if it stays)

realbigsean replied:
Had to dig into this again because it's been a while since I added that. I think we do need this.

If you check out the logs in some CI runs on [other branches](https://github.com/sigp/lighthouse/runs/7460369449?check_suite_focus=true) we never get past this log line, even though the `print_logs.sh` it looks like we should be dumping the whole file. 

`Jul 22 01:12:04.334 INFO Starting from known genesis state       service: beacon`

So I think what happens is we kick off all the processes and shut them down before they are fully initialized.

michaelsproul commented on Cargo.lock:
```diff
@@ -1839,6 +1930,28 @@ dependencies = [
+ "enr",
+ "hex",
+ "integer-sqrt",
+ "libp2p-core 0.32.1",
```
I've opened a PR upstream to see if we can remove this duplicate libp2p dependency: https://github.com/ralexstokes/ethereum-consensus/pull/139

michaelsproul replied:
With that PR merged I think a `cargo update` will sort this out, as `mev-rs` doesn't pin an `ethereum-consensus` version

realbigsean replied:
Looks like it worked! https://github.com/sigp/lighthouse/pull/3134/commits/5ad7632fc8481882b9367a118ae4b2f3c8c9d053#diff-13ee4b2252c9e516a0547f2891aa2105c3ca71c6d7a1e682c69be97998dfc87eL1944

michaelsproul replied:
Heck yeah!

realbigsean replied:
Ah dang there's actually a knock-on effect of this because `mev-build-rs` depends on `beacon-client-api` which used `PeerId`. I'm trying to update `beacon-client-api` here: https://github.com/ralexstokes/beacon-api-client/pull/28/files

realbigsean replied:
ok it looks all good now

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2900,7 +2902,7 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
                     BeaconSnapshot {
                         beacon_state: state,
-                        beacon_block: signed_block,
+                        beacon_block: signed_block.clone(),
```
I don't think we need this clone

```suggestion
                        beacon_block: signed_block,
```

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -4504,6 +4517,74 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+
+        // Check if finalization is advancing.
+        let current_epoch = current_slot.epoch(T::EthSpec::slots_per_epoch());
+        let epoch_since_finalization = current_epoch.saturating_sub(
```
```suggestion
        let epochs_since_finalization = current_epoch.saturating_sub(
```

realbigsean replied:
Added in ae1a500225471caac9c929daaf64195f3bea4b47

michaelsproul commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -24,6 +24,16 @@ pub struct ChainConfig {
     ///
     /// If set to 0 then block proposal will not wait for fork choice at all.
     pub fork_choice_before_proposal_timeout_ms: u64,
+    /// Number skip slots in a row before the BN refuses to use connected builders during payload construction.
```
```suggestion
    /// Number of skip slots in a row before the BN refuses to use connected builders during payload construction.
```

michaelsproul commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -24,6 +24,16 @@ pub struct ChainConfig {
     pub fork_choice_before_proposal_timeout_ms: u64,
+    /// Number skip slots in a row before the BN refuses to use connected builders during payload construction.
+    pub builder_fallback_skips: usize,
+    /// Number skip slots in the past `SLOTS_PER_EPOCH` before the BN refuses to use connected
```
```suggestion
    /// Number of skip slots in the past `SLOTS_PER_EPOCH` before the BN refuses to use connected
```

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -534,36 +567,101 @@ impl<T: EthSpec> ExecutionLayer<T> {
+
+                    return match (relay_result, local_result) {
+                        (Err(e), Ok(local)) => {
+                            warn!(self.log(),"Unable to retrieve a payload from a connected builder, falling back to the local execution client: {e:?}");
```
Do you mind wrapping this so it doesn't overflow 100 chars? Rustfmt usually gets confused with stuff like this

realbigsean replied:
Added line break related fixes here: 6093f992dbb4c6d313b1bad966298b5c4f172a6a

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -534,36 +567,101 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                            Ok(local)
+                        }
+                        (Ok(None), Ok(local)) => {
+                            warn!(self.log(), "No payload provided by connected builder. Attempting to propose through local execution engine");
```
Same here

michaelsproul commented on beacon_node/execution_layer/src/lib.rs:
```diff
@@ -534,36 +567,101 @@ impl<T: EthSpec> ExecutionLayer<T> {
+                                Ok(local)
+                            } else if !is_signature_valid {
+                                let pubkey_bytes = relay.data.message.pubkey;
+                                warn!(self.log(), "Invalid signature for pubkey {pubkey_bytes} on bid from connected builder, falling back to local execution engine.");
```
Same comment about wrapping for these logs :pray:

michaelsproul reviewed (commented):
I've reviewed about 1/3 of this. Will come back for more tomorrow

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

realbigsean committed:
- batch validator registration POST's
- Add an error log to the validator register endpoint so we can see exactly what the server returns

realbigsean committed:
Merge branch 'unstable' of https://github.com/sigp/lighthouse into pre-bellatrix-blinded-blocks

 Conflicts:
	beacon_node/http_api/src/lib.rs

realbigsean committed:
fix merge with unstable

realbigsean commented:
@michaelsproul  I added batching in the validator registration requests in a15a0d5123ca90434fcd27476cc3affe42add0e0 because I had issues with the relay when running 2.5k validators on ropsten.

michaelsproul commented on beacon_node/http_api/src/publish_blocks.rs:
```diff
@@ -0,0 +1,328 @@
+    block: SignedBeaconBlock<T::EthSpec, BlindedPayload<T::EthSpec>>,
+    log: Logger,
+) -> Result<SignedBeaconBlock<T::EthSpec, FullPayload<T::EthSpec>>, Rejection> {
+    let full_block = match block {
```
I think we could simplify this function a lot by using `try_into_full_block`, or the individual `From`/`Into` implementations for `Base` and `Altair` blocks.

https://github.com/sigp/lighthouse/blob/21dec6f603edd53ad8d2372941c77ab8098d3b5a/consensus/types/src/signed_beacon_block.rs#L310-L324

michaelsproul replied:
I was thinking something along the lines of

```rust
let full_payload = if let Ok(payload_header) = block.body().execution_payload() {
    // Logic to fetch full payload...
    Some(payload)
} else {
    None
};
block.try_into_full_block(full_payload)
```

realbigsean replied:
Love it! Added here: 78a1237a4aca3fab81203e3f443d4b111bcd9728

michaelsproul commented on book/src/builders.md:
```diff
@@ -0,0 +1,45 @@
+A primer on MEV can be found [here]([MEV](https://ethereum.org/en/developers/docs/mev/)).
+
+Lighthouse is able to interact with servers that implement the [builder API](https://github.com/ethereum/builder-specs). 
+If the builder you are connected to 
```
These docs are a bit incomplete. If we tidy them up to be minimally correct (e.g. no incomplete sentences or wrong info) I think we can merge now and come back for more comprehensive docs later

michaelsproul commented on common/eth2/src/lighthouse_vc/types.rs:
```diff
@@ -72,6 +92,8 @@ pub struct KeystoreValidatorsPostRequest {
     pub keystore: Keystore,
     pub graffiti: Option<GraffitiString>,
     pub suggested_fee_recipient: Option<Address>,
+    pub gas_limit: Option<u64>,
+    pub builder_proposals: Option<bool>,
```
I think we need `serde(default)` on these too?

realbigsean replied:
I think you're right, I added that and `serde(skip_serializing_if = "Option::is_none")` to all the optional fields in that request https://github.com/sigp/lighthouse/pull/3134/commits/5f6dc08ef3e16178e41868a8618adf5e816f11a9

michaelsproul commented on consensus/types/src/chain_spec.rs:
```diff
@@ -1355,4 +1355,12 @@ mod yaml_tests {
     }
+
+    #[test]
+    fn test() {
```
```suggestion
    fn test_domain_builder() {
```

michaelsproul commented on validator_client/src/block_service.rs:
```diff
@@ -269,7 +253,7 @@ impl<T: SlotClock + 'static, E: EthSpec> BlockService<T, E> {
                             },
                             Err(BlockError::Irrecoverable(e))  => {
-                                error!(log, "Error whilst producing a blinded block, cannot fallback because block was signed"; "error" => ?e);
+                                error!(log, "Error whilst producing a blinded block, cannot fallback because the block was signed"; "error" => ?e);
```
Could wrap this

realbigsean replied:
Added here: 4c67d7d625274c8c9eb9effc7f0405b3f5095c92

michaelsproul commented on validator_client/src/validator_store.rs:
```diff
@@ -379,6 +415,45 @@ impl<T: SlotClock + 'static, E: EthSpec> ValidatorStore<T, E> {
+    }
+
+    fn get_gas_limit_defaulting(&self, gas_limit: Option<u64>) -> u64 {
+        // If there is a `suggested_fee_recipient` in the validator definitions yaml
```
```suggestion
        // If there is a `gas_limit` in the validator definitions yaml
```

michaelsproul reviewed (changes requested):
Looking great on the whole, just a few minor comments and then we're ready to roll!

I've added the v2.5.0 tag too, but we can drop it from the release if we don't make the deadline

realbigsean committed:
Update beacon_node/beacon_chain/src/beacon_chain.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

realbigsean committed:
epochs_since_finalization typo

realbigsean committed:
Update beacon_node/beacon_chain/src/chain_config.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

realbigsean committed:
Update beacon_node/beacon_chain/src/chain_config.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

realbigsean committed:
Update consensus/types/src/chain_spec.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

realbigsean committed:
Update validator_client/src/validator_store.rs

Co-authored-by: Michael Sproul <micsproul@gmail.com>

realbigsean committed:
use  method during block reconstruction

realbigsean committed:
fix line breaks on logs and comments

realbigsean committed:
lockfile update

realbigsean committed:
fix some more log line lengths

michaelsproul reviewed (approved):
The review updates so far look good! Once you're happy with it I think we should MERGE 🐼

realbigsean committed:
make sure  skips serializing optional fields

realbigsean commented:
Working on docs and resolving this: https://github.com/sigp/lighthouse/pull/3134#discussion_r933348680 

But will merge when those things are all set!

realbigsean committed:
update lockfile

realbigsean committed:
docs

realbigsean commented:
Added docs and fixed the link check here https://github.com/sigp/lighthouse/pull/3134/commits/84e7f5375e3a41d41511593a7c17d1bc38a41488

Resolved the dependency issue here https://github.com/sigp/lighthouse/pull/3134/commits/eda5a5ca626da9042a28bc6e69d1c1132a8719d7

So I think we're good to go now :tada:

paulhauner committed:
Misc doc suggestions

realbigsean committed:
update cli help

realbigsean committed:
fix cli whitespace

paulhauner commented:
bors r+

paulhauner commented:
bors r-

paulhauner commented:
bors r+
